src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -261,3 +261,67 @@ function chol_blocked_rev(Σ̄::AbstractMatrix, L::AbstractMatrix, nb::Integer,
     # Convert to `Matrix`s because blas functions require StridedMatrix input.
     return chol_blocked_rev!(Matrix(Σ̄), Matrix(L), nb, upper)
 end
+
+#####
+##### `qr`
+#####
+
+const QR_TYPE = Union{QR, QRPivoted, LinearAlgebra.QRCompactWY}
+
+function rrule(::typeof(qr), X::AbstractMatrix{<:Real})
+    F = qr(X)
+
+    if size(F.R, 2) != size(F.R, 1)
+        throw(ArgumentError("Pullback for QR decomposition is only supported for m × n matrices with m >= n"))
+    end
+    function qr_pullback(Ȳ::Composite)
+        # `getproperty` on `Composite`s ensures we have no thunks.
+        ∂X = qr_rev(F, Ȳ.Q, Ȳ.R)
+        return (NO_FIELDS, ∂X)
+    end
+    return F, qr_pullback
+end
+
+function rrule(::typeof(getproperty), F::T, x::Symbol) where T <: QR_TYPE
+    function getproperty_qr_pullback(Ȳ)
+        C = Composite{T}
+        ∂F = if x === :Q
+            C(Q=Ȳ,)
+        elseif x === :R
+            C(R=Ȳ,)
+        end
+        return NO_FIELDS, ∂F, DoesNotExist()
+    end
+    if x === :Q
+        # Return thing Q for consistency
+        n = size(F.R, 1)
+        return F.Q[:, 1:n], getproperty_qr_pullback
```

Comment:
Is this on to do to the primal value?

Reply:
This was because the anything else does not contribute, and the tests used the result of this to infer dimensions. However after pondering on it I figured out it is probably better to let the `getproperty` pullback do just the obvious and have the `qr` pullback make sure to only take in account what is relevant and make sure the dimensions match.